### PR TITLE
Introduce LockedNORIV2 and test upgrade path.

### PR DESCRIPTION
**Pull request by scottynomad**
Follow-ups to address in another PR:
- [ ] Removal of `ContractNames` parameter from `getContractFactory` generic type.
- [ ] Use permit to streamline and re-introduce batch grant creation.
- [ ] Custom errors.

### CODE REVIEW

File: contracts/LockedNORIV2.sol
Comment (amiecorso) [inline contracts/LockedNORIV2.sol:226]: could probably make this language less specific now?

File: contracts/LockedNORIV2.sol
Comment (amiecorso) [inline contracts/LockedNORIV2.sol:220]: probably ok that none of this is indexed since we expect there to be very few of these events?

**Review by amiecorso — APPROVED** [review]

File: contracts/LockedNORIV2.sol
Comment (scottynomad) [reply contracts/LockedNORIV2.sol:220]: oh interesting, yeah I would think this is exceedingly rare and probably not of interest to indexers.

File: contracts/LockedNORIV2.sol
Comment (scottynomad) [reply contracts/LockedNORIV2.sol:226]: Good call I need to go back and look at these docs.

**Comment by scottynomad** [comment]
The LockedNORI tests are passing now.  The stacked branches make it difficult to see the changes made to `LockedNORI.sol` and `LockedNORI.test.ts`.  Once this merges down to jaycen's branch it should be somewhat easier to see.

File: contracts/LockedNORIV2.sol
Comment (jaycenhorton) [inline contracts/LockedNORIV2.sol:233]: is this upgrade a good opportunity to add custom errors?

File: contracts/test/ScheduleTestHarness.sol
Comment (jaycenhorton) [inline contracts/test/ScheduleTestHarness.sol:4]: Wonder if libraries can also be upgraded? could be nice to use [solidity global time units](https://docs.soliditylang.org/en/v0.8.15/units-and-global-variables.html#time-units)

File: contracts/test/LockedNORIV2Helper.sol
Comment (jaycenhorton) [inline contracts/test/LockedNORIV2Helper.sol:100]: :drooling_face:

File: contracts/test/DSTest/test.sol
Comment (jaycenhorton) [inline contracts/test/DSTest/test.sol:1]: was this actually needed? [This example repo](https://github.com/gakonst/foundry-rust-template/blob/master/contracts/src/test/Contract.t.sol) makes it look like you can just import this file using `import "ds-test/test.sol";`

The foundry book says it exists at a [different import](https://book.getfoundry.sh/forge/writing-tests.html?highlight=dstest#writing-tests) though

File: plugins/index.ts
Comment (jaycenhorton) [inline plugins/index.ts:60]: can we also include `localhost`?

File: test/LockedNORIUpgrade.test.ts
Comment (jaycenhorton) [inline test/LockedNORIUpgrade.test.ts:49]: missing await?

File: test/LockedNORIUpgrade.test.ts
Comment (jaycenhorton) [inline test/LockedNORIUpgrade.test.ts:72]: missing await?

File: test/LockedNORIUpgrade.test.ts
Comment (jaycenhorton) [inline test/LockedNORIUpgrade.test.ts:39]: missing await?

File: types/global.d.ts
Comment (jaycenhorton) [inline types/global.d.ts:199]: I can see how this generic can be frustrating to work with. I wrote this so to add type support and intellisense support when writing contracts deployments.


It also once helped with type inference when deploying contracts so you don't have to do `as ContractName` but that part appears to have broke (for some reason ContractFactory is actually causing a type error because it's not actually imported in global.d.ts and I think that's causing ts to infer a type of `any`)

Did you run  into issues with this? If it can't be reverted back to `ContractNames`, we should just delete this and stop trying to augment the types. 

(it's also possible the type augmentation i added here isn't required anymore and inference might be working by default in the latest types of various packages)

File: contracts/deprecated/ERC777PresetPausablePermissioned.sol
Comment (jaycenhorton) [inline contracts/deprecated/ERC777PresetPausablePermissioned.sol:1]: did this contract change at all?

File: contracts/LockedNORIV2.sol
Comment (jaycenhorton) [reply contracts/LockedNORIV2.sol:233]: I think you can actually just remove this function, unset `ERC777_TOKENS_RECIPIENT_HASH` in the 1820 registry and override any of they 777 recipient stuff if needed

File: contracts/LockedNORIV2.sol
Comment (jaycenhorton) [inline contracts/LockedNORIV2.sol:249]: do we want to use this instead of the permit pattern?

- using this approach: 2 serial transactions (approve + depositFor) (although it might be possible to only ever have to call `approve` once if we give it a high enough balance, erc 20 allows finfinite approvals even if the amount is > the current accounts balance)
- use erc 20 permit: 1 signature, and 1 transactions. requires figuring out typed data signing with fireblocks. Definitely safer, than infinite approvals, and fewer transactions (although signing would feel the same as a transaction in fireblocks since it still requires confirming in the app)

Comment (jaycenhorton) [reply contracts/LockedNORIV2.sol:249]: I think you'd actually have to also call `createGrant` using the non-permit pattern, right? 3 total transactions I think



`createGrant` -> `approve` -> `depositFor`? Is there a way to simplify that to two?



Permit could be done in 1 signature + 1 tx

File: contracts/LockedNORIV2.sol
Comment (jaycenhorton) [inline contracts/LockedNORIV2.sol:465]: perhaps `require(old !== new)`

File: test/LockedNORI.test.ts
Comment (jaycenhorton) [inline test/LockedNORI.test.ts:1]: https://www.diffchecker.com/R75nkGLJ

File: test/LockedNORI.test.ts
Comment (jaycenhorton) [inline test/LockedNORI.test.ts:377]: [await order is important for this](https://ethereum-waffle.readthedocs.io/en/latest/matchers.html#emitting-events)
```suggestion
          await expect(
            lNori
```

File: test/LockedNORI.test.ts
Comment (jaycenhorton) [inline test/LockedNORI.test.ts:419]: [await order is important for this](https://ethereum-waffle.readthedocs.io/en/latest/matchers.html#emitting-events)
```suggestion
          await expect(
            lNori
```

File: test/LockedNORI.test.ts
Comment (jaycenhorton) [inline test/LockedNORI.test.ts:774]:
```suggestion
    await expect(
      lNori
```

File: contracts/LockedNORIV2.sol
Comment (scottynomad) [reply contracts/LockedNORIV2.sol:233]: Good point, I removed the ERC777Recipient inheritance and this function so that sends will fail due to it not being present.  It makes sense to unregister the proxy from the 1820 registry as well.

File: contracts/test/ScheduleTestHarness.sol
Comment (scottynomad) [reply contracts/test/ScheduleTestHarness.sol:4]: I remember you pointed this out before but I wasn't clear where it was relevant since this library only deals in seconds anyway.

File: contracts/LockedNORIV2.sol
Comment (scottynomad) [reply contracts/LockedNORIV2.sol:176]: Sure, could do.  I figured it might be more self-explanatory / self-contained this way since we only interact with BridgedPolygonNORI as an ERC20 but I'm not strongly attached to that idea.

I'm aware that once I start fixing the vesting task we may want to add bulk grant creation and deposit functions in here which may change that story.

File: contracts/LockedNORIV2.sol
Comment (scottynomad) [reply contracts/LockedNORIV2.sol:249]: Yeah definitely intending to revisit this now that we know this upgrade is working cleanly.

File: test/LockedNORIUpgrade.test.ts
Comment (jaycenhorton) [inline test/LockedNORIUpgrade.test.ts:7]:
```suggestion
import { LockedNORI } from '.@/typechain-types/contracts/LockedNORI';
import LockedNORIArtifact from '@/legacy-artifacts/LockedNORI.sol/LockedNORI.json';
import { TestToken777 } from '@/typechain-types/contracts/test/TestToken777';
```

File: contracts/LockedNORIV2.sol
Comment (scottynomad) [reply contracts/LockedNORIV2.sol]: Will address custom errors into a later PR.

File: contracts/test/LockedNORIV2Helper.sol
Comment (scottynomad) [reply contracts/test/LockedNORIV2Helper.sol:73]: Nope, removed.

File: contracts/test/DSTest/test.sol
Comment (scottynomad) [reply contracts/test/DSTest/test.sol:1]: It is part of the foundry standard library but that's not landed yet.  I will plan to update the import once I get that merged in.

File: test/LockedNORIUpgrade.test.ts
Comment (scottynomad) [reply test/LockedNORIUpgrade.test.ts:4]: That seems to be how it's working. I'm going to remove this comment.

File: test/LockedNORIUpgrade.test.ts
Comment (scottynomad) [reply test/LockedNORIUpgrade.test.ts:39]: out of practice! :)

File: types/global.d.ts
Comment (scottynomad) [reply types/global.d.ts:200]: It had some trickle down effects with the renaming if I recall correctly.  Including `LockedNORI` as well as `LockedNORIV2` here causes it to clash with several other types that I'd like top leave `LockedNORI` out of.  Similar story for any solidity test contracts.  I will bookmark this to revisit when the dust settles as well.

File: contracts/deprecated/ERC777PresetPausablePermissioned.sol
Comment (scottynomad) [reply contracts/deprecated/ERC777PresetPausablePermissioned.sol:1]: Just the solidity version.

**Comment by scottynomad** [comment]
See #279

**Pull request closed without merging**
